src/main/java/com/khutircraftubackend/category/CategoryService.java

```diff
@@ -22,95 +21,112 @@
-
-            category.setParentCategory(parentCategory);
-        } else {
-            category.setParentCategory(null);
-        }
-        return categoryRepository.save(category);
-    }
-
-
-    @Transactional
-    public CategoryEntity updateCategory(Long id, CategoryUpdateRequest request, MultipartFile iconFile) throws IOException {
-        CategoryEntity existingCategory = findCategoryById(id);
-
-        if (iconFile != null && !iconFile.isEmpty()) {
-            existingCategory.setIconUrl(handleIcon(iconFile));
-        } else if (request.iconFile() != null) {
-            existingCategory.setIconUrl(handleIcon(iconFile));
-        }
-        categoryMapper.updateCategoryEntity(existingCategory, request);
-
-        if (request.parentCategoryId() != null) {
-            CategoryEntity parentCategory = findCategoryById(request.parentCategoryId());
-            existingCategory.setParentCategory(parentCategory);
-        } else {
-            existingCategory.setParentCategory(null);
-        }
-
-        return categoryRepository.save(existingCategory);
-    }
-
-    public void deleteCategory(Long id) {
-        deleteCategory(id, false);
-    }
-
-    @Transactional
-    public void deleteCategory(Long id, boolean forceDelete) {
-        List<CategoryEntity> childCategories = categoryRepository.findAllByParentCategory_Id(id);
-
-        if (childCategories.isEmpty()) {
-            categoryRepository.deleteById(id);
-        }
-
-
-        if (!childCategories.isEmpty() && forceDelete) {
-            for (CategoryEntity child : childCategories) {
-                deleteCategory(child.getId(), forceDelete);
-            }
-            categoryRepository.deleteById(id);
-        } else if(!childCategories.isEmpty()){
-            throw new CategoryDeletionException(CategoryExceptionMessages.CATEGORY_HAS_SUBCATEGORIES_OR_PRODUCTS);
-        }
-    }
+	
+	private final CategoryRepository categoryRepository;
+	private final CategoryMapper categoryMapper;
+	private final FileConverterService fileConverterService;
+	
+	private String handleIcon(MultipartFile iconFile) throws IOException {
+		
+		if (iconFile == null) {
+			return "";
+		}
+		
+		return fileConverterService.convert(iconFile);
+	}
+	
+	private CategoryEntity findCategoryById(Long id) {
+		
+		return categoryRepository.findById(id).orElseThrow(() ->
+				new CategoryNotFoundException(CategoryExceptionMessages.CATEGORY_NOT_FOUND));
+	}
+	
+	public CategoryEntity getCategoryById(Long id) {
+		
+		return findCategoryById(id);
+	}
+	
+	public List<CategoryResponse> getAllRootCategories() {
+		
+		return categoryRepository.findAllByParentCategoryIsNull().stream().map(
+				categoryMapper::toCategoryResponse
+		).collect(Collectors.toList());
+	}
+	
+	public List<CategoryResponse> getAllByParentCategoryId(Long id) {
+		
+		return categoryRepository.findAllByParentCategory_Id(id).stream().map(
+				categoryMapper::toCategoryResponse
+		).collect(Collectors.toList());
+	}
```

Comment:
Not very good idea to return something of type `CategoryResponse` from method in a Service. What if you need to use it elsewhere internally and need actual entity from the method in the Service? 

Better delegate "mapping" logic to controller.
Or, maybe it's a reasonable idea to implement a method in Mapper that accepts List of Entity objects and returns List of DTO objects? could you please google / research a bit ?

Reply:
Reworked.